go/tasks/plugins/k8s/ray/ray.go

```diff
@@ -0,0 +1,353 @@
+				},
+			},
+		},
+		{
+			Name: "MEMORY_LIMITS",
+			ValueFrom: &v1.EnvVarSource{
+				ResourceFieldRef: &v1.ResourceFieldSelector{
+					ContainerName: "ray-worker",
+					Resource:      "limits.cpu",
+				},
+			},
+		},
+		{
+			Name: "MY_POD_NAME",
+			ValueFrom: &v1.EnvVarSource{
+				FieldRef: &v1.ObjectFieldSelector{
+					FieldPath: "metadata.name",
+				},
+			},
+		},
+		{
+			Name: "MY_POD_IP",
+			ValueFrom: &v1.EnvVarSource{
+				FieldRef: &v1.ObjectFieldSelector{
+					FieldPath: "status.podIP",
+				},
+			},
+		},
+	}
+	primaryContainer.Env = append(primaryContainer.Env, podSpec.Containers[0].Env...)
+	primaryContainer.Lifecycle = &v1.Lifecycle{
+		PreStop: &v1.LifecycleHandler{
+			Exec: &v1.ExecAction{
+				Command: []string{
+					"/bin/sh", "-c", "ray stop",
+				},
+			},
+		},
+	}
+
+	primaryContainer.Ports = []v1.ContainerPort{
+		{
+			Name:          "redis",
+			ContainerPort: 6379,
+		},
+		{
+			Name:          "head",
+			ContainerPort: 10001,
+		},
+		{
+			Name:          "dashboard",
+			ContainerPort: 8265,
+		},
+	}
+
+	podTemplateSpec := v1.PodTemplateSpec{
+		Spec: v1.PodSpec{
+			Containers:     []v1.Container{*primaryContainer},
+			InitContainers: initContainers,
+		},
+	}
+	return podTemplateSpec
+}
+
+func (rayJobResourceHandler) BuildIdentityResource(ctx context.Context, taskCtx pluginsCore.TaskExecutionMetadata) (client.Object, error) {
+	return &rayv1alpha1.RayJob{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       KindRayJob,
+			APIVersion: rayv1alpha1.SchemeGroupVersion.String(),
+		},
+	}, nil
+}
+
+func getEventInfoForRayJob() (*pluginsCore.TaskInfo, error) {
+	taskLogs := make([]*core.TaskLog, 0, 3)
+	logPlugin, err := logs.InitializeLogPlugins(logs.GetLogConfig())
+
+	if err != nil {
+		return nil, err
+	}
+
+	if logPlugin == nil {
+		return nil, nil
+	}
+
+	// TODO: Retrieve the name of head pod from rayJob.status, and add it to task logs
+	// RayJob CRD does not include the name of the worker or head pod for now
+
+	// TODO: Add ray Dashboard URI to task logs
+
+	return &pluginsCore.TaskInfo{
+		Logs: taskLogs,
+	}, nil
+}
+
+func (rayJobResourceHandler) GetTaskPhase(ctx context.Context, pluginContext k8s.PluginContext, resource client.Object) (pluginsCore.PhaseInfo, error) {
+	rayJob := resource.(*rayv1alpha1.RayJob)
+	info, err := getEventInfoForRayJob()
+	if err != nil {
+		return pluginsCore.PhaseInfoUndefined, err
+	}
+	switch rayJob.Status.JobStatus {
+	case rayv1alpha1.JobStatusPending:
+		return pluginsCore.PhaseInfoNotReady(time.Now(), pluginsCore.DefaultPhaseVersion, "job is pending"), nil
+	case rayv1alpha1.JobStatusFailed:
+		reason := fmt.Sprintf("Failed to create Ray job: %s", rayJob.Name)
+		return pluginsCore.PhaseInfoFailure(errors.TaskFailedWithError, reason, info), nil
+	case rayv1alpha1.JobStatusSucceeded:
+		return pluginsCore.PhaseInfoSuccess(info), nil
+	}
+	return pluginsCore.PhaseInfoQueued(time.Now(), pluginsCore.DefaultPhaseVersion, "JobCreated"), nil
```

Comment:
Is the default case "Running" rather than "Queued"? Does the plugin ever report a "Running" phase?

Reply:
sorry, I missed it. Added running phase to it